client/components/bulk-select/index.jsx

```diff
@@ -21,33 +19,30 @@ export default React.createClass( {
 	},
 
 	getStateIcon() {
-		if ( this.hasSomeElementsSelected() ) {
+		if ( ! this.hasAllElementsSelected() && this.hasSomeElementsSelected() ) {
 			return <Gridicon className="bulk-select__some-checked-icon" icon="minus-small" size={ 18 }/>;
 		}
 	},
 
 	hasAllElementsSelected() {
-		return this.props.selectedElements && this.props.selectedElements === this.props.totalElements;
+		return this.props.selectedElements === this.props.totalElements;
 	},
 
 	hasSomeElementsSelected() {
-		return this.props.selectedElements && this.props.selectedElements < this.props.totalElements;
+		return this.props.selectedElements > 0;
 	},
 
 	handleToggleAll() {
-		const newCheckedState = ! ( this.hasSomeElementsSelected() || this.hasAllElementsSelected() );
-		this.props.onToggle( newCheckedState );
+		this.props.onToggle( this.hasSomeElementsSelected() );
```

Comment:
We probably should also refactor https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/plugins/plugin-list-header/index.jsx#L103 to use the received boolean instead of calculating it again.